### PR TITLE
Use ruff as a linter

### LINKED ISSUE #53: Use ruff
Change pre-commits to use ruff instead of flake8

### PR DESCRIPTION
**Pull request by noemifrisina**
Closes #53 . 
Replace isort/flake8 in pre-commits with Ruff.

Also pin builds to 3.11 as p4p doesn't build in 3.12 (stop docs from failing)

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -98,10 +98,10 @@
...
 # docs in the python documentation.
-intersphinx_mapping = dict(python=("https://docs.python.org/3/", None))
+intersphinx_mapping = {"python": ("https://docs.python.org/3/", None)}
 
 # A dictionary of graphviz graph attributes for inheritance diagrams.
-inheritance_graph_attrs = dict(rankdir="TB")
+inheritance_graph_attrs = {"rankdir": "TB"}
 
 # Common links that should be available on every page
 rst_epilog = """
```
Comment (DominicOram) [inline docs/conf.py:107]: Could: The links to `flake8` and `isort` links here aren't needed

**Review by DominicOram — CHANGES_REQUESTED** [review]
Great, thank you. Minor comment. Could you add https://github.com/DiamondLightSource/dodal/pull/221/commits/b980f9bf3df0e0042a650dd9d46540d12f7589ab in to this PR too so that the CI is working?

File: docs/conf.py (hunk 1 above)
Comment (noemifrisina) [reply docs/conf.py:107]: Done

**Comment by noemifrisina** [comment]
> Great, thank you. Minor comment. Could you add [DiamondLightSource/dodal@b980f9b](https://github.com/DiamondLightSource/dodal/commit/b980f9bf3df0e0042a650dd9d46540d12f7589ab) in to this PR too so that the CI is working?

Sure... I thought the fix was already on dodal main though? (I may be confusing it with another one)

**Review by dperl-dls — APPROVED** [review]

**Pull request merged**